Dbz 3459 add database history properties omitted from connector docs

roldanbob opened the pull request:
[DBZ-3459](https://issues.redhat.com/browse/DBZ-3459)

This change extracts the `database.history.*` properties from the tables of connector configuration properties for the Db2, MySQL, Oracle, and SQL Server connector docs  to a shared partial. 
It also adds consistent sections  to each of the docs to describe the use of passthrough properties for configuring database drivers and database history clients.

I tested the changes in local upstream and  downstream builds.

There are some other pending issues related to property tables, for example, [DBZ-3160](https://issues.redhat.com/browse/DBZ-3160), which concerns extracting the entirety of the connector properties into a shared source, and [DBZ-3120](https://issues.redhat.com/browse/DBZ-3120), which proposes to convert tables to description lists. Taking a stepwise approach, implementing single-sourcing for the database history properties first, and can later take on the full monty. Might eventually also want to move the pass-through sections to a shared file as well.

Please backport the changes to the 1.5 and 1.6 branches.

gunnarmorling commented on documentation/modules/ROOT/pages/connectors/db2.adoc:
```diff
@@ -2103,16 +2101,20 @@ The name format is _schema-name.table-name_.
- * All connector configuration properties that begin with the  `database.history.consumer.` prefix are used (without the prefix) when creating the Kafka consumer that reads the database history when the connector starts.
-
-For example, the following connector configuration properties {link-kafka-docs}.html#security_configclients[secure connections to the Kafka broker]:
+{prodname} relies on a Kafka producer to write schema changes to database history topics, and a Kafka consumer to read from database history topics when a connector starts.
```
The plural is a bit confusing here. The connector has exactly one database history topic.

gunnarmorling commented on documentation/modules/ROOT/pages/connectors/db2.adoc:
```diff
@@ -2130,9 +2132,19 @@ database.history.consumer.ssl.truststore.password=test1234
 ----
 
-Be sure to consult the {link-kafka-docs}.html[Kafka documentation] for all of the configuration properties for Kafka producers and consumers. Note that the Db2 connector uses the {link-kafka-docs}.html#newconsumerconfigs[new consumer].
+As is the case with the xref:{context}-pass-through-configuration-properties-for-database-drivers[pass-through properties for database drivers], {prodname} strips the prefixes from the properties before {prodname} it passes them to configure the database history producer or consumer client.
```
```suggestion
As is the case with the xref:{context}-pass-through-configuration-properties-for-database-drivers[pass-through properties for database drivers], {prodname} strips the prefixes from the properties before it passes them to configure the database history producer or consumer client.
```

gunnarmorling commented on documentation/modules/ROOT/pages/connectors/oracle.adoc:
```diff
@@ -1545,6 +1531,52 @@ By default, no operations are skipped.
+See the Kafka documentation for more details about link:https://kafka.apache.org/documentation.html#producerconfigs[Kafka producer configuration properties] and link:https://kafka.apache.org/documentation.html#consumerconfigs[Kafka consumer configuration properties].
+
+[id="{context}-database-history-properties-for-configuring-connector-behavior"]
+==== Database history properties for configuring connector behavior
```
I think it'd make more sense to discuss this before the DB history pass-through properties. Also, can't these introductory sentences go into that partials file?

gunnarmorling commented on documentation/modules/ROOT/pages/connectors/oracle.adoc:
```diff
@@ -1545,6 +1531,52 @@ By default, no operations are skipped.
 |===
 
+[id="{context}-pass-through-configuration-properties-for-database-drivers"]
+==== Pass-through configuration properties for database drivers
```
Can't this all go into the partials file?

gunnarmorling reviewed (commented):
@roldanbob, a few comments inline.

roldanbob commented:
@gunnarmorling Yes, the comments make sense. I was distracted by making sure that we took care of the dependencies related to the 1.5 backports, but I'll get back to this today. It was good for you to confirm that the passthrough sections that you suggest we include in the partials file are common to the four connectors. I wasn't 100% certain about that.

roldanbob commented:
@gunnarmorling Per the preceding conversation, I've incorporated your comments, and moved the pass-through sections into the shared file. Tested in a local Antora build. I think that this is good to go.

gunnarmorling commented:
@roldanbob, getting a few warnings when building this branch:

```
asciidoctor: WARNING: skipping reference to missing attribute: link-content-based-routing
asciidoctor: WARNING: skipping reference to missing attribute: link-filtering
asciidoctor: WARNING: skipping reference to missing attribute: routedbyvalue
asciidoctor: WARNING: skipping reference to missing attribute: routedbyvalue
asciidoctor: WARNING: skipping reference to missing attribute: routedbyvalue
asciidoctor: ERROR: db2.adoc: line 22: unmatched macro: endif::community[]
```

gunnarmorling commented on documentation/modules/ROOT/pages/connectors/oracle.adoc:
```diff
@@ -1545,6 +1531,56 @@ By default, no operations are skipped.
+
+The {prodname} connector provides for pass-through configuration of the database driver.
+Pass-through database properties begin with the prefix `database.*`.
+For example, the property `database.tinyInt1isBit=false` is passed to the JDBC URL.
```
tinyInt1isBit is specific to MySQL. Also, why isn't this part of a partial? I think, you'd want to have two partials:

* One for the DB pass through props
* One for the DB history stuff

roldanbob replied:
Same comment as earlier about commented out section. 

I already added the passthrough content to the DB history partial. I can split it out again into its own partial, but I'm not sure that's really necessary. WDYT?

gunnarmorling commented on documentation/modules/ROOT/pages/connectors/oracle.adoc:
```diff
@@ -1545,6 +1531,56 @@ By default, no operations are skipped.
+See the Kafka documentation for more details about link:https://kafka.apache.org/documentation.html#producerconfigs[Kafka producer configuration properties] and link:https://kafka.apache.org/documentation.html#consumerconfigs[Kafka consumer configuration properties].
+
+[id="{context}-passthrough-database-history-properties-for-configuring-producer-and-consumer-clients"]
+==== Pass-through database history properties for configuring producer and consumer clients
```
This should be part of the 2nd partial mentioned above.

roldanbob replied:
Ditto. This section no longer appears in the rendered doc, because it's commented out. I was waiting to prove out the downstream build before I remove it entirely.

gunnarmorling reviewed (commented):
@roldanbob, some more comments inline. I think there's still more duplication than there ought to be.

roldanbob commented:
@gunnarmorling 
> getting a few warnings when building this branch:

Hmm. Weird. I didn't touch the content-based routing or outbox event routing content, so the missing attribute errors are unrelated to any changes that I made. The two actual attributes that are mentioned are both defined in antora.yml, so not sure why they aren't found. If you were testing locally, do you have them defined in `playbook.yml`?

The other supposed missing attribute that Antora detected , `{routedByValue}` is not a documentation attribute at all, but a variable in the outbox configuration that's coded as `pass:[${routedByValue}]` in the text. I didn't make any changes there, and the same string appears in the `master` branch, so I don't understand why Antora has suddenly gotten more picky.  :\

roldanbob commented on documentation/modules/ROOT/pages/connectors/mysql.adoc:
```diff
@@ -2607,12 +2580,19 @@ database.history.consumer.ssl.truststore.password=test1234
+[id="{context}-database-history-properties-for-configuring-connector-behavior"]
+==== Database history properties for configuring connector behavior
+
+In addition to the pass-through database history properties that you use to configure producers or consumers, there is a set of `database.history.*` properties whose names do not include `producer` or `consumer`.
```
Sorry, this was confusing, because, per my earlier comment, I hadn't yet removed this section, but just commented it out, so it still shows in the diff.

roldanbob commented on documentation/modules/ROOT/partials/modules/all-connectors/ref-connector-configuration-database-history-properties.adoc:
```diff
@@ -0,0 +1,88 @@
+database.history.producer.ssl.truststore.location=/var/private/ssl/kafka.server.truststore.jks
+database.history.producer.ssl.truststore.password=test1234
+database.history.producer.ssl.key.password=test1234
+database.history.consumer.security.protocol=SSL
```
Done.

Naros commented:
@gunnarmorling @roldanbob those warnings are likely from older branches for situations where the `pass:[]` qualifier wasn't applied.  If you take a look at some outbox changes in 1.2 [here](https://raw.githubusercontent.com/debezium/debezium/1.2/documentation/modules/ROOT/pages/configuration/outbox-event-router.adoc) and you search for `routedByValue`, you'll notice some occurrences where the variable name isn't qualified by the `pass:[]` syntax but looking at 1.3 [here](https://raw.githubusercontent.com/debezium/debezium/1.3/documentation/modules/ROOT/pages/configuration/outbox-event-router.adoc), we can see those were fixed.  I suspect the same exists for the link situations as well.  The easiest way to find which branches are problematic are to remove the branches from the playbook and add them back one by one until you get warnings and then review the files on that branch.

roldanbob commented:
@gunnarmorling I did some re-jiggering that seems to resolve the downstream build errors that resulted from including the shared partial with the DB history properties, and I have otherwise incorporated the feedback from all of your comments.

Assuming that the explanation from @Naros about the build warnings satisfies you, I think that this is now ready to merge.

gunnarmorling commented:
> I didn't touch the content-based routing or outbox event routing content, so the missing attribute errors are unrelated to any changes that I made


> Assuming that the explanation from @Naros about the build warnings satisfies you

Yes, we can handle those separately. Generally speaking, I'd love for us to be a bit more pro-active here and take ownership. For instance, does it really matter whether something was not working correctly before? Why not quickly fix it in the context of such task. I understand a desire to focus on the change at hand and not be side-tracked, but still some proverbial picking up of litter along the way is always a good idea. It's really not about "who did it", but about producing the best outcome possible, right.

gunnarmorling commented on documentation/modules/ROOT/partials/modules/all-connectors/ref-connector-configuration-database-history-properties.adoc:
```diff
@@ -0,0 +1,98 @@
+
+* xref:{context}-database-history-properties-for-configuring-connector-behavior[Database history connector configuration properties] that control how {prodname} processes events that it reads from the database history topic.
+* xref:{context}-pass-through-database-history-properties-for-configuring-producer-and-consumer-clients[Pass-through database history properties] that control how Kafka producer and consumer clients interact with the Kafka broker.
+* xref:{context}-pass-through-configuration-properties-for-database-drivers[Pass-through database driver properties] that control the behavior of the database driver.
```
DB driver pass-through properties don't apply to the schema history. They apply to all the connectors. So this should be pulled up one level, to the same level as DB history properties, not be part of them.

roldanbob replied:
Ah, ok. I guess I do need to pull those into a separate _partial_ file. Downstream, I ran into build problems when I tried to insert heading levels into the partial, so I ended up using the simple `.` followed by no whitespace. But, of course, that means the 3 headings are all on the same level.

 OK, I'll take care of that. Thanks.

gunnarmorling commented on documentation/modules/ROOT/pages/connectors/mysql.adoc:
```diff
@@ -2120,12 +2121,14 @@ Information about the properties is organized as follows:
 * xref:mysql-required-connector-configuration-properties[Required connector configuration properties]
 * xref:mysql-advanced-connector-configuration-properties[Advanced connector configuration properties]
-* xref:mysql-pass-through-configuration-properties[Pass-through configuration properties]
+* xref:debezium-{context}-connector-database-history-and-pass-through-configuration-properties[{prodname} connector database history and pass-through configuration properties]
```
See my other comment below; DB drive pass-through properties should be their own section, on the same level as the history ones.

gunnarmorling reviewed (commented):
@roldanbob, something is still not quite right in terms of the overall organization/grouping.

roldanbob commented:
> ... Generally speaking, I'd love for us to be a bit more pro-active here and take ownership. For instance, does it really matter whether something was not working correctly before?

All that I meant was that I don't understand where the errors originate. AFAIK, we didn't see the attributes errors earlier, and I don't see them in my local build.

gunnarmorling commented:
> All that I meant was that I don't understand where the errors originate. AFAIK, we didn't see the attributes errors earlier, and I don't see them in my local build.

No worries. I sensed a bit of "not my department" spirit in the original reply, and I don't think it's what we should aspire for. Re the warnings, I'm getting them when building upstream/1.2, so they have been around for a while. For the Db2 one, I already pushed a fix to upstream/1.4 to remove the superfluous tag.

gunnarmorling commented:
Btw. the PR will need a rebase (plus force push) to address the conflict in the Oracle connector docs page.

gunnarmorling commented:
Ok, pushed another commit to upstream/1.2 to address the remaining warnings in the Antora build. All good on that front now.

roldanbob commented:
@gunnarmorling  
> Btw. the PR will need a rebase (plus force push) to address the conflict in the Oracle connector docs page.
Ok, demonstrating my ignorance here. I've not done this before. To rebase for the PR, do I rebase this branch from `master`?

gunnarmorling commented:
> To rebase for the PR, do I rebase this branch from master?

You'd do something like this:

    git checkout master
    git pull upstream master
    git
checkout DBZ-3459-add-database-history-properties-omitted-from-connector-docs
    git rebase master
    // resolve any potential merge conflicts, then do git rebase --continue
    git push -f
origin DBZ-3459-add-database-history-properties-omitted-from-connector-docs

>

roldanbob commented:
Well that took more doing than I would have liked. I had a challenge resolving that `%$#&` merge conflict. But I think I've got it sorted now, and once again I'm suggesting (with appropriate caution and crossed fingers) that this is ready to merge.

gunnarmorling commented on documentation/modules/ROOT/partials/modules/all-connectors/ref-connector-configuration-database-history-properties.adoc:
```diff
@@ -0,0 +1,78 @@
+The safe default is `false`.
+Skipping should be used only with care as it can lead to data loss or mangling when the binlog is being processed.
+
+|[[{context}-property-database-history-store-only-monitored-tables-ddl]]<<{context}-property-database-history-store-only-monitored-tables-ddl, `+database.history.store.only.monitored.tables.ddl+`>> +
```
```suggestion
|[[{context}-property-database-history-store-only-monitored-tables-ddl]]<<{context}-property-database-history-store-only-monitored-tables-ddl, `+database.history.store.only.monitored.tables.ddl+`>>
```

gunnarmorling commented on documentation/modules/ROOT/partials/modules/all-connectors/ref-connector-configuration-database-history-properties.adoc:
```diff
@@ -0,0 +1,78 @@
++
+_Deprecated and scheduled for removal in a future release; use xref:{context}-property-database-history-store-only-captured-tables-ddl[`database.history.store.only.captured.tables.ddl`] instead._
+|`false`
+|A Boolean value that specifies whether the connector should record all DDL statements  +
```
```suggestion
|A Boolean value that specifies whether the connector should record all DDL statements 
```

gunnarmorling commented on documentation/modules/ROOT/partials/modules/all-connectors/ref-connector-configuration-database-history-properties.adoc:
```diff
@@ -0,0 +1,78 @@
+|`false`
+|A Boolean value that specifies whether the connector should record all DDL statements  +
++
+`true` records only those DDL statements that are relevant to tables whose changes are being captured by {prodname}. Set to `true` with care because missing data might become necessary if you change which tables have their changes captured. +
```
```suggestion
`true` records only those DDL statements that are relevant to tables whose changes are being captured by {prodname}. Set to `true` with care because missing data might become necessary if you change which tables have their changes captured.
```

gunnarmorling commented on documentation/modules/ROOT/partials/modules/all-connectors/ref-connector-configuration-database-history-properties.adoc:
```diff
@@ -0,0 +1,78 @@
+
+|[[{context}-property-database-history-store-only-captured-tables-ddl]]<<{context}-property-database-history-store-only-captured-tables-ddl, `+database.history.store.only.captured.tables.ddl+`>>
+|`false`
+|A Boolean value that specifies whether the connector should record all DDL statements  +
```
```suggestion
|A Boolean value that specifies whether the connector should record all DDL statements
```

gunnarmorling commented on documentation/modules/ROOT/partials/modules/all-connectors/ref-connector-configuration-database-history-properties.adoc:
```diff
@@ -0,0 +1,78 @@
+|`false`
+|A Boolean value that specifies whether the connector should record all DDL statements  +
++
+`true` records only those DDL statements that are relevant to tables whose changes are being captured by {prodname}. Set to `true` with care because missing data might become necessary if you change which tables have their changes captured. +
```
```suggestion
`true` records only those DDL statements that are relevant to tables whose changes are being captured by {prodname}. Set to `true` with care because missing data might become necessary if you change which tables have their changes captured.
```

gunnarmorling commented on documentation/modules/ROOT/partials/modules/all-connectors/ref-connector-pass-through-database-driver-configuration-properties.adoc:
```diff
@@ -0,0 +1,5 @@
+Pass-through database properties begin with the prefix `database.*`.
+For example, the connector passes properties such as `database.foobar=false` to the JDBC URL.
+
+As is the case with the xref:{context}-pass-through-database-history-properties-for-configuring-producer-and-consumer-clients[pass-through properties for database history clients], {prodname} strips the prefixes from the properties before it passes them to configure the database history producer or consumer client.
```
```suggestion
As is the case with the xref:{context}-pass-through-database-history-properties-for-configuring-producer-and-consumer-clients[pass-through properties for database history clients], {prodname} strips the prefixes from the properties before it passes them to the database driver.
```

!!!

roldanbob replied:
Good catch. Neglected to update the sentence after I moved it from the other section. Thanks!

gunnarmorling commented on documentation/modules/ROOT/pages/connectors/oracle.adoc:
```diff
@@ -1519,6 +1517,14 @@ If the captured table(s) schema changes infrequently or never, this is the ideal
 |The number of hours in the past from SYSDATE to mine archive logs.
 Using the default `0` will mine all archive logs.
 
+|[[oracle-property-log-mining-archive-log-only-mode]]<<oracle-property-log-mining-archive-log-only-mode, `+log.mining.archive.log.only.mode+`>>
+|`false`
+|Controls whether or not the connector mines changes from just archive logs or a combination of the online redo logs and archive logs (the default). +
+ +
+For some environments where online redo logs are archived frequently enough to cause LogMiner session failures,
+this opt-in feature only mines archive logs which are guaranteed to be reliable unlike the circular buffer nature of redo logs that can be archived at any point, leading to LogMiner session errors.
+Note that by using this opt-in feature, events will be emitted with a certain amount of latency that is entirely dependent on how frequently online redo logs are archived by Oracle.
```
Where is this one coming from?

roldanbob replied:
This property is in the main properties table in the version of `oracle.adoc` on `master`.

Naros replied:
This was added in [DBZ-3561](https://issues.redhat.com/browse/DBZ-3561) to satisfy a situation where redo logs were being moved to archive faster than the connector could keep up and mine them and since the redo logs were named within sequences in their names, i.e. `redoa.log`, `redob.log`, etc that the connector would fail more often than not just due to the speed at which logs rotated.

gunnarmorling replied:
Ok, gotcha. Why is it facing up in this PR, though?

gunnarmorling commented on documentation/modules/ROOT/partials/modules/all-connectors/ref-connector-configuration-database-history-properties.adoc:
```diff
@@ -0,0 +1,78 @@
+
+|[[{context}-property-database-history-kafka-bootstrap-servers]]<<{context}-property-database-history-kafka-bootstrap-servers, `+database.history.kafka.bootstrap.servers+`>>
+|
+|A list of host/port pairs that the connector uses for establishing an initial connection to the Kafka cluster. This connection is used for retrieving database schema history previously stored by the connector, and for writing each DDL statement read from the source database. Each pair should point to the same Kafka cluster used by the Kafka Connect process.
```
```suggestion
|A list of host/port pairs that the connector uses for establishing an initial connection to the Kafka cluster. This connection is used for retrieving the database schema history previously stored by the connector, and for writing each DDL statement read from the source database. Each pair should point to the same Kafka cluster used by the Kafka Connect process.
```

roldanbob replied:
It's in db2 (L2892), mysql (L2174), and sqlserver (L2006) docs in the `master` branch. Should it not be included?

gunnarmorling replied:
I'm not sure what you're asking? I just found that the "the" is missing there (see darker green marked area).

roldanbob replied:
@gunnarmorling Sorry, I responded in the wrong place. I was answering your question about where did this one come from: https://github.com/debezium/debezium/pull/2455#discussion_r657830855

gunnarmorling reviewed (commented):
@roldanbob, ok, I think we're getting there. Some (hopefully) last suggestions after another review. Note you can "Add suggestion to batch" and "Commit batch" these directly from the UI.

roldanbob commented on documentation/modules/ROOT/partials/modules/all-connectors/ref-connector-configuration-database-history-properties.adoc:
```diff
@@ -0,0 +1,78 @@
+{prodname} provides a set of `database.history.*` properties that control how the connector processes events that it reads from the schema history topic.
+These `database.history` properties specify how the {context} connector processes the event data that it reads from the database history.
+
+The following table describes the `database.history` properties for configuring the {prodname} {context} connector.
```
Thanks for catching that. You're suggesting that we create a new variable for the connector name/context, along the lines e.g., of `{connector-name-context}`? That would is fine with me, but I'm not sure whether we need the extra overhead, unless there are lots of instances where it would provide added value over the simpler solution, which in this case means removing the connector name, so that the sentence reads _"... specify how the connector processes ..."_ WDYT?

gunnarmorling replied:
Yes, we can do that, too of course. In fact, it is what I will do now right away, so to get this one merged finally. You could have done yourself, you're the docs lead and should decide such matters :)

Naros commented:
> I don't see them in my local build.

@roldanbob, I don't believe you would see them in your build as you said you had defined a variable in the playbook or antora configuration file for downstream.  In upstream, we don't do that because `${routedByValue}` isn't a variable, its really just a simple circumstance where the syntax we use overlaps with some Antora functionality and the correct way to fix the older branches would be for us to simply push commits to them to add the `pass:[]` brackets, that's all.

> No worries. I sensed a bit of "not my department" spirit in the original reply,

@gunnarmorling My apologies if that was from me, that was certainly not the intent.

gunnarmorling commented:
Squashed and applied to master (=1.6 and 1.5). Thanks, @roldanbob!

Taking a step back, this really took a long time. Not sure what we can do better here, but I wished we'd handle these changes a tad more efficient. The price we pay here is literally quite high.
